torch/nn/functional.py

```diff
@@ -5173,21 +5173,26 @@ def multi_head_attention_forward(
     # (deep breath) calculate attention and out projection
     #
 
-    B, Nt, E = q.shape
-    q_scaled = q / math.sqrt(E)
     if attn_mask is not None:
-        attn_output_weights = torch.baddbmm(attn_mask, q_scaled, k.transpose(-2, -1))
-    else:
-        attn_output_weights = torch.bmm(q_scaled, k.transpose(-2, -1))
-    attn_output_weights = softmax(attn_output_weights, dim=-1)
-    if dropout_p > 0.0:
-        attn_output_weights = dropout(attn_output_weights, p=dropout_p)
+        if attn_mask.size(0) == 1:
+            attn_mask = attn_mask.unsqueeze(0)
+        else:
+            attn_mask = attn_mask.view(bsz, num_heads, -1, src_len)
+
+    q = q.view(bsz, num_heads, -1, head_dim)
+    k = k.view(bsz, num_heads, -1, head_dim)
+    v = v.view(bsz, num_heads, -1, head_dim)
+
+    attn_output_2, attn_output_weights_2 = _scaled_dot_product_attention(
+        q, k, v, attn_mask, dropout_p, need_weights, False)
+    attn_output_2 = attn_output_2.transpose(1, 2).transpose(0, 1).contiguous().view(bsz * tgt_len, embed_dim)
```

Comment:
The transpose(0 calls necessary to pull this off might cost us significant overheads.

There's an impedeance mismatch not just between functional.MHA and sdp_attention, but also between nn.MHA and functional.MHA.

nn.MHA treats batch_first as preferred format, but functional.MHA appears to only support not batch_first. I see two options:
1 - pass in the bacth_first variable (but that's somewhat non-preferred because we might end up with control flow as a function of an input, the input being batch_first.)
2 - have an inverted polarity MHA which prefers batch_first and use that one internally, then have nn.MHA and legacy compatibility functional.MHA call the new implementation - with a possible path to deprecating the old interface eventually)